### PR TITLE
283 - tables for report data

### PR DESCRIPTION
**Pull request by pbastia**
PR for: https://github.com/bcgov/cas-reporting/issues/283

### CODE REVIEW

**Review by joshgamache — APPROVED** [review]
Looks great, tests all passing locally. 

Reading through the models, everything is matching up against the architecture outlined in the ticket, great work!

File: bc_obps/reporting/migrations/00010_report_data_tables.py [hunk 1]
```diff
@@ -0,0 +1,692 @@
...
+                        to='registration.user',
+                    ),
+                ),
+            ],
+            options={
+                'db_table': 'erc"."report_activity',
+                'db_table_comment': 'A table to store the reported activity-specific data, in a JSON format',
+            },
+        ),
+        migrations.CreateModel(
```
Comment (dleard) [inline bc_obps/reporting/migrations/00010_report_data_tables.py:308]: I don't see the fkeys to the unit / source type data in the migration here (though they are defined in the model?)

File: bc_obps/reporting/migrations/00010_report_data_tables.py [hunk 2]
```diff
@@ -0,0 +1,692 @@
...
+                        to='registration.user',
+                    ),
+                ),
+            ],
+            options={
+                'db_table': 'erc"."report_fuel',
+                'db_table_comment': 'A table to store the reported activity-specific data, in a JSON format',
+            },
+        ),
+        migrations.CreateModel(
```
Comment (dleard) [inline bc_obps/reporting/migrations/00010_report_data_tables.py:373]: I don't see the fkey to the source type data object (though it is defined in the model?)

File: bc_obps/reporting/models/report_emission.py [hunk 3]
```diff
@@ -0,0 +1,33 @@
...
+    report_fuel = models.ForeignKey(
+        ReportFuel,
+        null=True,
+        on_delete=models.CASCADE,
+        related_name="%(class)s_records",
+        db_comment="The fuel data this emission data belongs to, if applicable",
+    )
+
+    class Meta:
+        db_table_comment = "A table to store the reported activity-specific data, in a JSON format"
```
Comment (dleard) [inline bc_obps/reporting/models/report_emission.py:31]:
```suggestion
        db_table_comment = "A table to store the reported emission-specific data, in a JSON format"
```

File: bc_obps/reporting/models/report_fuel.py [hunk 4]
```diff
@@ -0,0 +1,33 @@
...
+    )
+    fuel_type = models.ForeignKey(
+        FuelType,
+        on_delete=models.PROTECT,
+        related_name="%(class)s_records",
+        db_comment="The fuel type this data applies to",
+    )
+
+    class Meta:
+        db_table_comment = "A table to store the reported activity-specific data, in a JSON format"
```
Comment (dleard) [inline bc_obps/reporting/models/report_fuel.py:31]:
```suggestion
        db_table_comment = "A table to store the reported fuel-specific data, in a JSON format"
```

File: bc_obps/reporting/models/report_methodology.py [hunk 5]
```diff
@@ -0,0 +1,18 @@
...
+
+    report_emission = models.OneToOneField(
+        ReportEmission,
+        on_delete=models.CASCADE,
+        related_name="report_methodology",
+        db_comment="The emission data this methodology applies to",
+    )
+
+    class Meta:
+        db_table_comment = "A table to store the reported activity-specific data, in a JSON format"
```
Comment (dleard) [inline bc_obps/reporting/models/report_methodology.py:16]:
```suggestion
        db_table_comment = "A table to store the reported methodology-specific data, in a JSON format"
```

File: bc_obps/reporting/models/report_source_type.py [hunk 6]
```diff
@@ -0,0 +1,31 @@
...
+    )
+    report_activity = models.ForeignKey(
+        ReportActivity,
+        on_delete=models.CASCADE,
+        related_name="%(class)s_records",
+        db_comment="The activity data record this source type data belongs to",
+    )
+
+    class Meta:
+        db_table_comment = "A table to store the reported activity-specific data, in a JSON format"
```
Comment (dleard) [inline bc_obps/reporting/models/report_source_type.py:29]:
```suggestion
        db_table_comment = "A table to store the reported source type-specific data, in a JSON format"
```

File: bc_obps/reporting/models/report_unit.py [hunk 7]
```diff
@@ -0,0 +1,18 @@
...
+
+    report_source_type = models.ForeignKey(
+        ReportSourceType,
+        on_delete=models.CASCADE,
+        related_name="%(class)s_records",
+        db_comment="The source type data this unit data belongs to",
+    )
+
+    class Meta:
+        db_table_comment = "A table to store the reported activity-specific data, in a JSON format"
```
Comment (dleard) [inline bc_obps/reporting/models/report_unit.py:16]:
```suggestion
        db_table_comment = "A table to store the reported unit-specific data, in a JSON format"
```

File: bc_obps/reporting/models/report_unit.py [hunk 8]
```diff
@@ -0,0 +1,18 @@
+from django.db import models
+from reporting.models.report_data_base_model import ReportDataBaseModel
+from reporting.models.report_source_type import ReportSourceType
+
+
+class ReportUnit(ReportDataBaseModel):
```
Comment (dleard) [inline bc_obps/reporting/models/report_unit.py:6]: When we talk about handling reports for multiple facilities, we probably want to consider whether all data models should have a fkey to that faciltiy_report for ease of flattening the data without needing a lot of transitive joins.

**Review by dleard — APPROVED** [review]
A couple nits on the table comments & I don't see some of the fkeys that are defined in the models showing up in the migration file, unless I somehow missed them or there's a reason they don't show up?

File: bc_obps/reporting/migrations/00010_report_data_tables.py (hunk 1 above)
Comment (pbastia) [reply bc_obps/reporting/migrations/00010_report_data_tables.py:308]: I think it's because of how django creates the models, the "report source type" model doesn't exist when this code is run yet. Once the model is added later on, we can add the fkeys (see Line 596 and 681). 

TLDR; the fields are there, just further down the file

File: bc_obps/reporting/migrations/00010_report_data_tables.py (hunk 2 above)
Comment (pbastia) [reply bc_obps/reporting/migrations/00010_report_data_tables.py:373]: same as earlier, it's added on Line 606

File: bc_obps/reporting/models/report_unit.py (hunk 8 above)
Comment (pbastia) [reply bc_obps/reporting/models/report_unit.py:6]: We've added a fkey to the report version everywhere already, I wonder if that one would be useful too? I'll create a card for that

Comment (pbastia) [reply bc_obps/reporting/models/report_unit.py:6]: https://github.com/bcgov/cas-reporting/issues/304
